cluster-autoscaler/cloudprovider/openstack/openstack_nodegroup.go

```diff
@@ -0,0 +1,288 @@
+func (ng *OpenstackNodeGroup) DeleteNodes(nodes []*apiv1.Node) error {
+
+	// Batch simultaneous deletes on individual nodes
+	ng.nodesToDeleteMutex.Lock()
+	// Check that these nodes would not make the batch delete more nodes than the minimum would allow
+	if *ng.targetSize-len(ng.nodesToDelete)-len(nodes) < ng.MinSize() {
+		ng.nodesToDeleteMutex.Unlock()
+		return fmt.Errorf("deleting nodes would take nodegroup below minimum size")
+	}
+	ng.nodesToDelete = append(ng.nodesToDelete, nodes...)
+	ng.nodesToDeleteMutex.Unlock()
+
+	ng.clusterUpdateMutex.Lock()
+	defer ng.clusterUpdateMutex.Unlock()
+
+	ng.nodesToDeleteMutex.Lock()
+	if len(ng.nodesToDelete) == 0 {
+		// Deletion was handled by another goroutine
+		ng.nodesToDeleteMutex.Unlock()
+		return nil
+	}
+	ng.nodesToDeleteMutex.Unlock()
+
+	// This goroutine has the clusterUpdateMutex, so will be the one
+	// to actually delete the nodes. While this goroutine waits, others
+	// will add their nodes to nodesToDelete and block at acquiring
+	// the clusterUpdateMutex lock. One they get it, the deletion will
+	// already be done and they will return above at the check
+	// for len(ng.nodesToDelete) == 0.
+	time.Sleep(deleteNodesBatchingDelay)
+
+	ng.nodesToDeleteMutex.Lock()
+	nodes = make([]*apiv1.Node, len(ng.nodesToDelete))
+	copy(nodes, ng.nodesToDelete)
+	ng.nodesToDelete = nil
+	ng.nodesToDeleteMutex.Unlock()
+
+	var nodeNames []string
+	for _, node := range nodes {
+		nodeNames = append(nodeNames, node.Name)
+	}
+	klog.V(1).Infof("Deleting nodes: %v", nodeNames)
+
+	updatePossible, currentStatus, err := ng.openstackManager.CanUpdate()
+	if err != nil {
+		return fmt.Errorf("could not check if cluster is ready to delete nodes: %v", err)
+	}
+	if !updatePossible {
+		return fmt.Errorf("can not delete nodes, cluster is in %s status", currentStatus)
+	}
+
+	minionCount, err := ng.openstackManager.NodeGroupSize(ng.id)
+	if err != nil {
+		return fmt.Errorf("could not get current node count")
+	}
+
+	if minionCount-len(nodes) < ng.MinSize() {
+		return fmt.Errorf("size decrease too large, desired:%d min:%d", minionCount-len(nodes), ng.MinSize())
+	}
+
+	var minionIPs []string
+	for _, node := range nodes {
+		for _, addr := range node.Status.Addresses {
+			if addr.Type == apiv1.NodeInternalIP {
+				minionIPs = append(minionIPs, addr.Address)
+				break
+			}
+		}
+	}
+
+	err = ng.openstackManager.DeleteNodes(ng.id, minionIPs, minionCount-len(nodes))
+	if err != nil {
+		return fmt.Errorf("manager error deleting nodes: %v", err)
+	}
+
+	// Block until cluster has gone into update status and then completed
+	err = ng.WaitForClusterStatus(clusterStatusUpdateInProgress, waitForUpdateStatusTimeout)
+	if err != nil {
+		return fmt.Errorf("scale down error: %v", err)
+	}
+	err = ng.WaitForClusterStatus(clusterStatusUpdateComplete, waitForCompleteStatusTimout)
+	if err != nil {
+		return fmt.Errorf("scale down error: %v", err)
+	}
+
+	*ng.targetSize, err = ng.openstackManager.NodeGroupSize(ng.id)
```

Comment:
Nodes are deleted through Heat and then Magnum is updated afterward?  The wait-for-update-to-finish makes a bit more sense now :).  I hope that we could go all through Heat or all through Magnum.  I know that the scale down in Magnum does not currently support node specification, so while that is being worked on, maybe Heat could be used exclusively for scale up and scale down and have Magnum update itself?